### PR TITLE
sql: remove implicit casts

### LINKED ISSUE #2071: sql: implicit casts sometimes allowed
to my knowledge, the sites are:
- [ ] concat operator `||` accepts non-strings: https://github.com/cockroachdb/cockroach/blob/master/sql/parser/eval.go#L244:L264
- [ ] `CONCAT` accepts non-strings (after #2055 is merged): https://github.com/cockroachdb/cockroach/pull/2055/files#diff-74eeff841a0488dbfb9b804729361666R137
- [x] integer division returns a float (postgres does integer division, and so should we): https://github.com/cockroachdb/cockroach/blob/master/sql/parser/eval.go#L226:L228

### PR DESCRIPTION
**Pull request by tamird**
Fixes #2071.

### CODE REVIEW

File: sql/parser/builtins.go [hunk 1]
```diff
@@ -457,26 +459,9 @@ func stringBuiltin3(f func(string, string, string) (Datum, error), returnType Da
...
-			return "t", nil
-		}
-		return "f", nil
-
-	case dNull:
-		return "", nil
-
-	default:
-		return "", fmt.Errorf("argument type unsupported: %s", datum.Type())
+	if dString, ok := datum.(DString); ok {
```
Comment (vivekmenezes) [inline sql/parser/builtins.go]: Is there any value to this function any more?

Comment (tamird) [inline sql/parser/builtins.go]: just not having to write the error message twice

**Comment by vivekmenezes** [comment]
LGTM

**Comment by petermattis** [comment]
LGTM

**Pull request merged**